src/core/thread/mle_router.cpp

```diff
@@ -3675,6 +3678,21 @@ void MleRouter::SendAddressSolicitResponse(const Coap::Message    &aRequest,
 
     Log(kMessageSend, kTypeAddressReply, aMessageInfo.GetPeerAddr());
 
+    // If assigning a new RLOC16 (e.g., on promotion of a child to
+    // router role) we clear any address cache entries associated
+    // with the old RLOC16.
+
+    if ((aResponseStatus == ThreadStatusTlv::kSuccess) && (aRouter != nullptr))
```

Comment:
We don't have confirmation that the Address Solicit Response actually made it back to the child here. On one hand, it is not fatal since we'll just trigger another Address Query. On the other hand, it the reason the Address Solicit Response may have been dropped may be due to congestion and sending a multicast Address Query is probably not going to help. Thoughts?

Reply:
Yes. Nice! I was thinking the same :) I convinced myself that this situation (Address Solicit being dropped and never reaching the child) should not cause any harm, as you said we end up doing additional address query (and it should be sent behind address solicit, and will be before it in the queue).